### ISSUE TITLE
Average Precision for each class

### ISSUE THREAD

**Issue by julian-douglas** [ultralytics/yolov5#4357]
## Is there any way to see the average precision for each class?

I have run my model for 1,000 epochs and I have a bunch of metrics (which are AMAZING by the way, thanks so making it so easy to see them!) and I have mAP, but I was wondering if there was a way to see the AP for each class? Like a table or something. 

In addition, is it possible to see the precision-recall graphs for each class? I can see something in the images tab on wandb, but as I have 80 classes, it looks very messy.

**Comment by akashAD98** [comment]
@glenn-jocher is there any way to calculate this ??

**Comment by mxy5201314** [comment]
This may require you to write the relevant code yourself

**Comment by glenn-jocher** [comment]
@julian-douglas @akashAD98 @mxy5201314 per-class metrics are displayed automatically by val.py for datasets with < 50 classes. You can force this manually with 
```
python val.py --verbose
```

**Comment by julian-douglas** [comment]
Hi @glenn-jocher , 

This is very helpful, thank you! However, when training had finished, a mAP of 0.6809 was recorded, however, when I run val.py --verbose --data '../data.yaml' --weights 'best.pt', I get a mAP of 0.591 across all classes, even when I changed the validation set directory to train/images. What is the reason for this disparity?

**Comment by glenn-jocher** [comment]
@julian-douglas 👋 hi, thanks for letting us know about this problem with YOLOv5 🚀. We've created a few short guidelines below to help users provide what we need in order to get started investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

* ✅ **Minimal** – Use as little code as possible that still produces the same problem
* ✅ **Complete** – Provide **all** parts someone else needs to reproduce your problem in the question itself
* ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

In addition to the above requirements, for [Ultralytics](https://ultralytics.com/) to provide assistance your code should be:

* ✅ **Current** – Verify that your code is up-to-date with current GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been resolved by previous commits.
* ✅ **Unmodified** – Your problem must be reproducible without any modifications to the codebase in this repository. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all of the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) and providing a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by julian-douglas** [comment]
Panic over! I was using the wrong image size 🤦🏼‍♂️

**Comment by glenn-jocher** [comment]
@julian-douglas yes you should train and val/detect at the same size for best results.
